lib/omniauth/protect/validator.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module Omniauth
+  module Protect
+    class Validator
+      def initialize(env, encoded_masked_token)
+        @session = env['rack.session']
+        @encoded_masked_token = encoded_masked_token
+      end
+
+      def valid_csrf_token?
+        begin
+          masked_token = Base64.urlsafe_decode64(@encoded_masked_token)
+        rescue ArgumentError # @encoded_masked_token is invalid Base64
+          return false
+        end
+
+        token_length = ActionController::RequestForgeryProtection::AUTHENTICITY_TOKEN_LENGTH
+
+        if masked_token.length == token_length * 2
+          csrf_token = unmask_token(masked_token, token_length)
+
+          real_token = real_csrf_token(token_length)
+          global_token = global_csrf_token(real_token)
```

Comment:
This is new code that's not just a refactor of the existing code, the only new problems this could introduce are timing attacks on L26 - is that a problem? :thinking: 

No because if you have the real token you're in already, and if you don't then everything takes the same time so no new information is leaked.

Reply:
🦆